src/components/ContactsAdd.js

```diff
@@ -1,39 +1,119 @@
-import { useState } from "react"
+import { useState } from "react";
 import { useNavigate } from "react-router-dom";
 
 function ContactsAdd(props) {
+  const initialForm = {
+    firstName: "",
+    lastName: "",
+    street: "",
+    city: "",
+  };
 
   // setContacts and contacts must be passed as props
   // to this component so new contacts can be added to the
   // state
-  const { setContacts, contacts } = props
+  const { setContacts, contacts } = props;
+
+  const [firstName, setFirstName] = useState("");
+  const [lastName, setLastName] = useState("");
+  const [street, setStreet] = useState("");
+  const [city, setCity] = useState("");
 
   //TODO: Implement controlled form
   //send POST to json server on form submit
 
+  const onFirstNameChange = (e) => {
+    setFirstName(e.target.value);
+  };
+  const onLastNameChange = (e) => {
+    setLastName(e.target.value);
+  };
+  const onStreetChange = (e) => {
```

Comment:
This is nice Jason!

As a suggestion, you have used a dynamic value to set each of the fields (e.target.value), so perhaps you could do the same thing for the field? 

Something like this 

const handleChange = (e) => {
    setAddressData({ ...addressData, [e.target.name]: e.target.value })
  }